### ISSUE TITLE
Git autodeploy some time does not trigger

### ISSUE THREAD

**Issue by oak-tree** [projectkudu/kudu#2117]
We have now two slots : production and staging. Only the staging connected to git - bitbucket. 
What we are trying to achieve is the following :

1 . push code to repo 
2. the staging will automaticly detect a change and pull changes 
3. it will stop tomcat server
4. pull the changes 
5. rerun tomcat
6. when its ready do a swap

It seems that for 3,4,5 we just go for touch web.config. But for some reason not all of the time the wepbapp is aware of change in the repo. I cannot put my finger on what is the cause but sometimes it capture repo update and sometimes is not

Any ideas how this subject?
Thanks

**Comment by davidebbo** [comment]
If the deployment doesn't get triggered at all, is it fair to say that everything in your steps relating to tomcat and later swapping the slots (i.e. everything after step 2) is not relevant to the question?

**Comment by oak-tree** [comment]
I Pushed  21 minute ago to `bitbucket`
![image](https://cloud.githubusercontent.com/assets/667153/18056167/cb4a63e4-6dd8-11e6-8a76-cafe41b146d2.png)

But in azure 
![image](https://cloud.githubusercontent.com/assets/667153/18056145/b6949eb0-6dd8-11e6-99f7-c115a1cd642c.png)

**Comment by davidebbo** [comment]
Try looking at the WebHook history on the bitbucket side. It should show if it was sent. And on Kudu side look under `D:\home\LogFiles\kudu\trace` to see if there was any activity.

**Comment by oak-tree** [comment]
In `bitbucket` repo there is no `WebHook` installed but there are `services` related to `azure`. It should do `post` to the webapp

> Services allow you to extend what Bitbucket does when the repository changes (for example, new code is pushed or a pull request is merged).

But they(bitbucket) do say that "webhooks will replace these services"

**Comment by davidebbo** [comment]
Yes, sorry, I meant Services. Though it seems they changed their UI and no longer show history.

Did you check the Kudu logs?

**Comment by oak-tree** [comment]
Just did, it weird. It does get the post - but it always one commit behind. If I do `git pull` in the `kudu` `repo` it fetches the last commit. 
```
Found this between the logs. maybe can help 


<step title="Incoming Request" date="2016-08-29T21:35:38.649" instance="...." url="/deploy?scmType=BitbucketGit" method="POST" ...... >
    <step title="Cleanup Xml Logs" date="2016-08-29T21:35:38.680" />
    <!-- duration: 48ms -->
    <step title="FetchHandler" date="2016-08-29T21:35:38.789" >
        <step title="Scm is not enabled, reject all requests." date="2016-08-29T21:35:38.946" />
        <!-- duration: 20ms -->
    </step>
    <!-- duration: 176ms -->
    <step title="Outgoing response" date="2016-08-29T21:35:38.980" type="response" statusCode="403" statusText="Forbidden" Server="Microsoft-IIS/8.0" Cache-Control="private" X-AspNet-Version="4.0.30319" />
    <!-- duration: 13ms -->
</step>
<!-- duration: 344ms -->
```

**Comment by oak-tree** [comment]
If I do `git fetch --dry-run` in `kudu repo` I see there is a fetch to be done

**Comment by oak-tree** [comment]
I tried to remove `bitbucket` and reset it. It does trigger on deploy. at least the `staging` slots is restarting. But it does **not** fetch the lastest version

**Comment by davidebbo** [comment]
And you're sure you're hooping up the right branch?

Can you share your web app name, either directly or [indirectly](https://github.com/projectkudu/kudu/wiki/Reporting-your-site-name-without-posting-it-publicly)? This will help us investigate. Thanks!

**Comment by oak-tree** [comment]
Sure, here is a dummy name :) http://dummya.azurewebsites.net/
and about the branch we have `azure` branch. Moreover and `get fetch --dry-run` in the repo shows there is a fetch available

**Comment by davidebbo** [comment]
There are two other sites there, including one with a slot. Is it the one with '3' in the name or the other?

**Comment by oak-tree** [comment]
The one that ends with `3`

**Comment by davidebbo** [comment]
If you look at the Kudu trace under `D:\home\LogFiles\kudu\trace`, you can see the full sequence of git commands that Kudu executes. With some reformatting (and I tokenized the org/repo name to not reveal it):

```
git remote rm origin
git remote add -t azure origin 'git@bitbucket.org:{yourorg}/{yourrepo}.git'
git branch
git fetch origin --progress
git update-ref refs/heads/azure origin/azure
git checkout azure --force
```

I don't want to affect your Kudu repo, but can you try going to `D:\home\site\repository` and running these commands yourself? Does that result in the same issue? If so, fo you see anything wrong with the steps that would cause this?

**Comment by oak-tree** [comment]
sure, give me a mintue

**Comment by oak-tree** [comment]
The `git branch` should include the `branch` does not have any parameter?

**Comment by davidebbo** [comment]
No, that's ok. It's just a call we make to test what the current branch is. I could have left it off the sequence, but I just listed every git invocation that was made.

**Comment by oak-tree** [comment]
ok, I ran them. Those commands run ok. slow, but ok

**Comment by davidebbo** [comment]
By 'run ok', you mean that they got the git head to the right commit? If so, it's very puzzling, as it's exactly what we run when the bitbucket hook fires.

**Comment by oak-tree** [comment]
Yes, It gets the right commit. Now the `site/repo` folder is newer than the `webapp` folder in one commit

**Comment by oak-tree** [comment]
## OK, I did some digging:

-  I deleted the `logs` at `trace` folder 
- I started our `deployment process` - We have our own built deployment process
- I checked what happens with the logs 

## Here are the results: 
In our deployment system, we always have an extra tag for quick rollback, we called it ` last_development`

The tag is being deleted and recreate each time a new deploy process is starting. The tags also being push to remote. 

So here is what happend, for some reason the push of the `remove/add` tag triggers the `deployment` cycle at `kudu`, Since this is one heavy java/tomcat/hibernate app it takes it time to boot. Meanwhile  our development process has finished to run compile + run tests and it pushes the update to repo. This `triggers` another `deployment cycle` at `kudu`. But, this get reject (forbidden) I guess is due the fact that the system is still down


What do you think? What would pushing `tags` triggers the `bitbucket services`?

**Comment by oak-tree** [comment]
I see that `bitbuckets` has two options 

1. use `service` wheneven there is a change in repo, this makes some kind of sense that pushing `tags` triggers the `service`
2. using `webhook` this allows more customization when to trigger the event

Seems like `service` will be deprecated in the near future, moreover it makes sense to use `webhook` only when the `branch` has changes and not on every repo change, doesn't it?

**Comment by davidebbo** [comment]
Still, I don't quite understand the behavior. What I would have expected is:

- You push your tag
- bb fires the hook
- Kudu sees that it's not for the relevant branch, and makes it a no-op
- You push the branch and normal things happen

But here, it seems that the tag push is triggering a full deployment, which I don't understand. We need to get to the bottom of that.

**Comment by watashiSHUN** [comment]
could you provide me with the `git` commands in your `deployment process`
Especially the ones interact with the remote repository in their exact order (ie)
```
git push origin 
git push origin --tags
```
that will help me to reproduce the issue

**Comment by oak-tree** [comment]
@davidebbo 
It seems like `kudu` does not understand this is not relevent for the brnach and trigger full deployment

@watashiSHUN, here are the list of remote commands

> #delete last fast recovery tag 
> git tag -d last_development || true
> git push --delete origin last_development || true
> 
> #create last_development tag for quick recovery
> git tag  -a last_development -m "quick recovery tag"
> git push --tags
> 
> #push current branch changes to remote
> git push origin master

Note that the `azure deployment cycle` starts before the last command so the last command fails. To be sure, I switched the order and do moved all `tags stuff` to the end. This make things work as expected.

**Comment by watashiSHUN** [comment]
@oak-tree 

as you mentioned above, "I switched the order and do moved all `tags stuff` to the end. This make things work as expected.", this just proves that kudu will **not** trigger development because of the `tag` deletion and recreation.

what happened here is that when kudu receives the notification about `git push --delete origin last_development`, it ran a series `git command` to check if there's an update in the gitbucket repo 

```
git remote rm origin
git remote add -t azure origin 'git@bitbucket.org:{yourorg}/{yourrepo}.git'
git branch
git fetch origin --progress
git update-ref refs/heads/azure origin/azure
git checkout azure --force
```

In most cases, this **should be a no-op**, however, since we ran `git push origin master` right after `git tag updates`,the gitbucket repo will already have the tags updated **and the newest commits** while kudu is still processing the `tag removal`. This is why you see that kudu triggers the deployment right after it receives the first gitbucket request

**while I am working on a fix, you can add a custom delay in your script as a workaround**
```
::delete last fast recovery tag 
git tag -d last_development
git push --delete origin last_development

::create last_development tag for quick recovery
git tag -a last_development -m "quick recovery tag"
git push --tags

:: this quick and dirty fix should keep your workflow unchanged
timeout 5

git push origin master
```

**Comment by watashiSHUN** [comment]
@oak-tree (I have updated my comment)

as you mentioned above, "I switched the order and do moved all `tags stuff` to the end. This make things work as expected.", this just proves that kudu will **not** trigger development because of the `tag` deletion and recreation.

what happened here is that when kudu receives the notification about `git push --delete origin last_development`, it ran a series `git command` to check if there's an update in the gitbucket repo 

```
git remote rm origin
git remote add -t azure origin 'git@bitbucket.org:{yourorg}/{yourrepo}.git'
git branch
git fetch origin --progress
git update-ref refs/heads/azure origin/azure
git checkout azure --force
```

In most cases, this **should be a no-op**, however, since we ran `git push origin master` right after `git tag updates`,the gitbucket repo will already have the tags updated **and the newest commits** before kudu finish processing `git fetch` for the `tag removal request`. This is why you see that kudu triggers the deployment "prematurely"

However. this should not cause your website to be 1 commit behind... after you finished running your git script, can you go to kudu and check for the commit ID?

it should be under `D:\home\site\deployments\` folder, where you would find a list of `commit hashes`, try to compare the most recent one against your gitbucket repository

**Comment by watashiSHUN** [comment]
@oak-tree 
you are seeing one commit behind because of the **auto swap**

here is what happened

1. you ran your script of git commands
2. gitbucket sends to kudu notifications about `tag delete, create, git push`
3. kudu process it **one by one**
4. kudu `fetch` remote repo and realized that there are new commits from `git push`
5. kudu thus starts an **azure deployment cycle**
6. after the deployment, azure performs an **auto swap**
7. after the swap, **changes from the latest deployment is moved to your production site and the staging site now looks like the production site before deployment(which is one commit behind)**

step 4 and step 7 combined makes people think that "kudu performs a **redeployment** on `tag delete`"

Could you verify that after deployment, your production site is ahead of the staging site by one commit?
you can check for commit hash in `D:\home\site\deployments\active` file of their scm site

**Comment by siroky** [comment]
Hi, we were confused in a similar way in case of staging slot connected to github and autoswap to production slot which wasn't connected anywhere. From the azure portal, it's actually not possible to see (or at least I didn't find it out) what's currently deployed on production slot. The web app section "Deployments" is visible only on the staging slot, but not on production, where it offers us to configure deployment source:

![image](https://cloud.githubusercontent.com/assets/435787/18402758/50a0b76c-76e1-11e6-9d53-93fc7dd32116.png)

I guess it would make sense in the Azure Portal, to split it into two sections:
- One could be "Deployment Source" and it would allow the users to configure the deployment source or disconnect it.
- The other one "Deployments" would show the deployments no matter if there is actually any deployment source connected at the moment.

As of now, e.g. if we have to see the deployment log, we have to go to KUDU, check the deployments JSON, find the the log url there and visit it. I understand that this request should be better suited in Azure User Voice, but wanted to share it here anyway.
